crates/evm/src/precompile_cache.rs

```diff
@@ -0,0 +1,341 @@
+/// `ContextStatefulPrecompileMut`. They are explicitly banned.
+#[derive(Debug)]
+#[cfg(feature = "std")]
+pub struct PrecompileCache {
+    /// Cache for precompile results and gas bounds.
+    cache: Cache<PrecompileKey, CacheEntry>,
+}
+
+#[cfg(feature = "std")]
+impl Default for PrecompileCache {
+    fn default() -> Self {
+        Self {
+            cache: mini_moka::sync::CacheBuilder::new(100_000)
+                .build_with_hasher(alloy_primitives::map::DefaultHashBuilder::default()),
+        }
+    }
+}
+
+/// Metrics for the cached precompile provider, showing hits / misses for each cache
+#[cfg(feature = "metrics")]
+#[derive(reth_metrics::Metrics, Clone)]
+#[metrics(scope = "sync.caching")]
+pub(crate) struct CachedPrecompileMetrics {
+    /// Precompile cache hits
+    precompile_cache_hits: metrics::Gauge,
+
+    /// Precompile cache misses
+    precompile_cache_misses: metrics::Gauge,
+
+    /// Precompile cache size
+    ///
+    /// NOTE: this uses the moka caches' `entry_count`, NOT the `weighted_size` method to calculate
+    /// size.
+    precompile_cache_size: metrics::Gauge,
+}
+
+/// A custom precompile provider that wraps a precompile provider and potentially a cache for it.
+#[derive(Clone, Debug)]
+pub struct MaybeCachedPrecompileProvider<P> {
+    /// The precompile provider to wrap.
+    precompile_provider: P,
+    /// The cache to use.
+    #[cfg(feature = "std")]
+    cache: Option<alloc::sync::Arc<PrecompileCache>>,
+    /// The spec id to use.
+    spec: SpecId,
+    /// Cache metrics.
+    #[cfg(feature = "metrics")]
+    metrics: CachedPrecompileMetrics,
+}
+
+impl<P> MaybeCachedPrecompileProvider<P> {
+    /// Given a [`PrecompileProvider`]  and cache for a specific precompile provider,
+    /// create a cached wrapper that can be used inside Evm.
+    #[cfg(feature = "std")]
+    pub fn new_with_cache(
+        precompile_provider: P,
+        cache: alloc::sync::Arc<PrecompileCache>,
+    ) -> Self {
+        Self {
+            precompile_provider,
+            #[cfg(feature = "std")]
+            cache: Some(cache),
+            spec: Default::default(),
+            #[cfg(feature = "metrics")]
+            metrics: Default::default(),
+        }
+    }
+
+    /// Creates a new `MaybeCachedPrecompileProvider` with cache disabled.
+    pub fn new_without_cache(precompile_provider: P) -> Self {
+        Self {
+            precompile_provider,
+            #[cfg(feature = "std")]
+            cache: None,
+            spec: Default::default(),
+            #[cfg(feature = "metrics")]
+            metrics: Default::default(),
+        }
+    }
+}
+
+impl<CTX: ContextTr, P: PrecompileProvider<CTX, Output = InterpreterResult>> PrecompileProvider<CTX>
+    for MaybeCachedPrecompileProvider<P>
+{
+    type Output = P::Output;
+
+    fn set_spec(&mut self, spec: <CTX::Cfg as Cfg>::Spec) -> bool {
+        self.precompile_provider.set_spec(spec.clone());
+        self.spec = spec.into();
+        true
```

Comment:
not sure if always returning true is right here, but I also find this return type a bit confusing

Reply:
good point, according to https://github.com/bluealloy/revm/blob/35a0611281997608dc9d1bd655b348105334f451/crates/handler/src/precompile_provider.rs#L15-L18 it should only return true if the spec id was changed, fixed.